apps/api-documenter/src/documenters/YamlDocumenter.ts

```diff
@@ -677,11 +650,46 @@ export class YamlDocumenter {
 
   private _getYamlItemName(apiItem: ApiItem): string {
     if (apiItem.parent && apiItem.parent.kind === ApiItemKind.Namespace) {
-      // For members a namespace, show the full name excluding the package part:
-      // Example: excel.Excel.Binding --> Excel.Binding
-      return this._getUid(apiItem).replace(/^[^.]+\./, '');
+      // If the immediate parent is a namespace, then add the namespaces to the name.  For example:
+      //
+      //   // Name: "N1"
+      //   export namespace N1 {
+      //     // Name: "N1.N2"
+      //     export namespace N2 {
+      //       // Name: "N1.N2.f(x,y)"
+      //       export function f(x: string, y: string): string {
+      //         return x + y;
+      //       }
+      //
+      //
+      //       // Name: "N1.N2.C"
+      //       export class C {
+      //         // Name: "member(x,y)"  <===========
+      //         public member(x: string, y: string): string {
+      //           return x + y;
+      //         }
+      //       }
+      //     }
+      //   }
+      //
+      // In the above example, "member(x, y)" does not appear as "N1.N2.C.member(x,y)" because YamlDocumenter
+      // embeds this entry in the web page for "N1.N2.C", so the container is obvious.  Whereas "N1.N2.f(x,y)"
+      // needs to be qualified because the DocFX template doesn't make pages for namespaces.  Instead, they get
+      // flattened into the package's page.
+      const nameParts: string[] = [ Utilities.getConciseSignature(apiItem) ];
+
+      for (let current: ApiItem | undefined = apiItem.parent; current; current = current.parent) {
+        if (current.kind !== ApiItemKind.Namespace) {
+          break;
+        }
+
+        nameParts.unshift(current.displayName);
+      }
+
+      return nameParts.join('.');
```

Comment:
If the `DeclarationReference` already is namespace qualified, could you leverage that? I'm thinking of adding some additional utility methods/properties to the `DeclarationReference` API to make that even easier.

Reply:
This is not really a `DeclarationReference`. It's a title that is shown in the document, in a specific context.

>  I'm thinking of adding some additional utility methods/properties to the `DeclarationReference` API to make that even easier.

@rbuckton  In order to integrate this into the TSDoc parser, we will need to integrate it into [NodeParser](https://github.com/microsoft/tsdoc/blob/cecc67a3c0590afc3f550de44e9d32a4da6c28cc/tsdoc/src/parser/NodeParser.ts#L921), and the objects need to replace the earlier [DocDeclarationReference](https://github.com/microsoft/tsdoc/blob/cecc67a3c0590afc3f550de44e9d32a4da6c28cc/tsdoc/src/nodes/DocDeclarationReference.ts#L34) AST node type.

Ideally we would do that in a way that eliminates the need for a standalone `DeclarationReference` API.  We can follow up about this separately.

Reply:
No worries, was just thinking about whether the DeclarationReference could be reused for this because it already supports the dot-qualified formatting, but I wouldn't worry about changing anything here.

Reply:
My thought had been to add a few members to DeclRef and related objects:
- `isFullyQualified` - `true` if the DeclRef has no missing components (`:function`, `:constructor`, `:index`, `:call`, `:new`, and `:complex` would also require an overload index), otherwise `false`
- `isReachable` - `true` if the DeclRef contains no Local (`~`) navigations, otherwise `false`.
- `isStatic` - `true` if the DeclRef contains only Static (`.`) navigations, otherwise `false`.
- `isInstance` - `true` if the DeclRef contains any Instance (`#`) navigations, otherwise `false`.
- `namespaceQualifiedName` - Returns the string value of the DeclRef's symbol excluding any meaning or overload index (i.e. `my-package!a.b.c:member` -> `a.b.c`, `my-package!a.b#c:member` -> `a.b#c`).
  - NOTE: I considered replacing `#` with `.prototype.` for this as well...
- `namespace` - Returns the string value of the DeclRef's symbol excluding the outermost ComponentPaths that do not represent a namespace member, meaning, or overload index (i.e. `my-package:a.b.c:member` -> `a.b`).
  - NOTE: This would have been a good use case for `enummember`, as an `enummember` is kind of `member`-like and kind of a `var`-like export of its containing namespace...
  - `my-package:a.b.c:member` -> `a` (members can't belong to namespaces (except enums, see above))
  - `my-package:a.b.c:class` -> `a.b` (classes can belong to namespaces)
  - `my-package:a.b.c:var` -> `a.b`
- `classQualifiedName` - Similar to `namespaceQualifiedName`, but only includes the containing class name for members:
  - `my-package:a.b.c:member` -> `b.c`
  - `my-package:a.b#c:member` -> b#c`
  - `my-package:a.b.c:var` -> ''
  - `my-package:a.b:constructor` -> `b`

Then you could have done something like:
```ts
const namespaceName = apiItem.canonicalReference
return apiItem.canonicalReference.namespaceQualifiedName + Utilities.getConsiseSignature(apiItem);
```

Reply:
> * `my-package:a.b.c:member` -> `b.c`
> * `my-package:a.b#c:member` -> `b#c`

It might be nice to have a standardized API for rendering names, especially if they may contain problematic characters like ECMAScript symbols, quoted strings, or overloads.

However, I'm unsure whether the `DeclarationReference` API should try to solve a problem that is really about documentation formatting.  For example, although `#` and `(2)` are important operators in the `{@link}` notation, I'm not sure they will look nice on the documentation website.  `Utilities.getConciseSignature()` returns `(arg1,arg2)` instead of the overload index because it's a better documentation experience, despite being a worse design for the UID requirements.

If we have `my-package!myNamespace.MyClass#staticMember(2)`, the kinds of strings we'd want from a doc formatting API might be more like this:
- `myNamespace (namespace)`
- `myNamespace.MyClass (class)` 
- `MyClass (class)`
- `staticMember(arg1,arg2)`
- `MyClass.staticMember(arg1,arg2) (static)`

The **api-extractor-model** API might be a better place to implement this, since it has access to information like whether the parent is a namespace or not, and what are the argument names.